common/config_parser.hpp

```diff
@@ -0,0 +1,161 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <string>
+#include <vector>
+#include <common/error.hpp>
+#include <common/expected.hpp>
+
+#ifndef MENDER_COMMON_CONFIG_PARSER_HPP
+#define MENDER_COMMON_CONFIG_PARSER_HPP
+
+namespace mender::common::config_parser {
+
+using namespace std;
+
+/** HttpsClient holds the configuration for the client side mTLS
+	configuration */
+struct HttpsClient {
+	string certificate;
+	string key;
+	string ssl_engine;
+};
+
+/** Security structure holds the configuration for the client Added for MEN-3924
+	in order to provide a way to specify PKI params outside HttpsClient. */
+struct ClientSecurity {
+	string auth_private_key;
+	string ssl_engine;
+};
+
+/** Connectivity instructs the client how we want to treat the keep alive
+	connections and when a connection is considered idle and therefore closed */
+struct ClientConnectivity {
+	bool disable_keep_alive = false;
+	int idle_conn_timeout_seconds = 0;
+};
+
+enum ConfigParserErrorCode {
+	ParseError,
+};
+
+using ConfigParserError = mender::common::error::Error<ConfigParserErrorCode>;
+using ExpectedBool = mender::common::expected::Expected<bool, ConfigParserError>;
+
+class MenderConfigFromFile {
+public:
+	/** Path to the public key used to verify signed updates.
+		Only one of artifact_verify_key/artifact_verify_keys can be specified. */
+	string artifact_verify_key;
+
+	/** List of verification keys for verifying signed updates.
+		Starting in order from the first key in the list,
+		each key will try to verify the artifact until one succeeds.
+		Only one of artifact_verify_key/artifact_verify_keys can be specified. */
+	vector<string> artifact_verify_keys;
```

Comment:
In Golang we had to keep both members because the struct is directly used in the parsing. However, in C++ I'm wondering if we should omit the singleton and use only the list. We can just populate it with one element if the singleton exists in the conf file, and it should make the rest of the logic simpler.

Reply:
To me it makes more sense to have them separate in this class because then it reflects what the file contained. `MenderConfig` could have them merged, but then it couldn't be based on this (`MenderConfigFromFile`) class which would be quite unfortunate.